lib/health/health.go

```diff
@@ -122,32 +132,59 @@ func (aggregator *HealthAggregator) Summary() *HealthReport {
 	aggregator.mutex.Lock()
 	defer aggregator.mutex.Unlock()
 
+	var failedLivenessChecks []string
+	var failedReadinessChecks []string
+
 	// In the absence of any reporters, default to indicating that we are both live and ready.
 	summary := &HealthReport{Live: true, Ready: true}
 
 	// Now for each reporter...
 	for name, reporter := range aggregator.reporters {
-		log.WithFields(log.Fields{
-			"name":           name,
-			"reporter-state": reporter,
-		}).Debug("Detailed health state")
-
 		// Reset Live to false if that reporter is registered to report liveness and hasn't
 		// recently said that it is live.
-		if summary.Live && reporter.reports.Live && (!reporter.latest.Live ||
-			(reporter.timeout != 0 && time.Since(reporter.timestamp) > reporter.timeout)) {
+		stillLive := reporter.latest.Live && !reporter.TimedOut()
+		if summary.Live && reporter.reports.Live && !stillLive {
 			summary.Live = false
 		}
 
 		// Reset Ready to false if that reporter is registered to report readiness and
 		// hasn't recently said that it is ready.
-		if summary.Ready && reporter.reports.Ready && (!reporter.latest.Ready ||
-			(reporter.timeout != 0 && time.Since(reporter.timestamp) > reporter.timeout)) {
+		stillReady := reporter.latest.Ready && !reporter.TimedOut()
+		if summary.Ready && reporter.reports.Ready && !stillReady {
 			summary.Ready = false
 		}
+
+		switch {
+		case reporter.reports.Live && !stillLive:
+			failedLivenessChecks = append(failedLivenessChecks, name)
+		case reporter.reports.Ready && !stillReady:
```

Comment:
This case can only be triggered if the one above is not so a reporter that is neither live nor ready will only show up as not live.  I guess that's OK since liveness is more severe than readiness but for debug purposes, might be better to have both bits of information.